File: MaskedEditText/src/main/java/com/vicmikhailau/maskededittext/IFormattedString.java
package com.vicmikhailau.maskededittext;


public interface IFormattedString extends CharSequence  {
    String getUnMaskedString();
}


abstract class AbstractFormattedString implements IFormattedString{
    private String mFormattedString;
    private final String mUnmaskedString;
    final Mask mMask;


    AbstractFormattedString(Mask mask, String rawString){
        mMask = mask;
        mUnmaskedString = buildRawString(rawString);
    }


    abstract String formatString();
    abstract String buildRawString(String str);

    public String getUnMaskedString() {
        return mUnmaskedString;
    }


    @Override
    public int length() {
        return toString().length();
    }

    @Override
    public String toString() {
        if (mFormattedString == null){
            mFormattedString = formatString();
        }
        return mFormattedString;
    }

    @Override
    public CharSequence subSequence(int start, int end) {
        return toString().subSequence(start, end);
    }

    @Override
    public char charAt(int index) {
        return toString().charAt(index);
    }
}

class FormattedString extends AbstractFormattedString {

    FormattedString(Mask mask, String rawString) {
        super(mask, rawString);
    }

    String buildRawString(String str) {
        StringBuilder builder = new StringBuilder();
        int inputLen = Math.min(mMask.size(), str.length());
        for (int i = 0; i < inputLen; i++){
            char ch = str.charAt(i);
            if (!mMask.isValidPrepopulateCharacter(ch))
                builder.append(ch);
        }
        return builder.toString();
    }

    String formatString() {
        StringBuilder builder = new StringBuilder();

        int strIndex = 0;
        int maskCharIndex = 0;
        char stringCharacter;

        while (strIndex < getUnMaskedString().length() && maskCharIndex < mMask.size()) {
            MaskCharacter maskChar = mMask.get(maskCharIndex);

            stringCharacter = getUnMaskedString().charAt(strIndex);

            if (maskChar.isValidCharacter(stringCharacter)) {
                builder.append(maskChar.processCharacter(stringCharacter));
                strIndex += 1;
                maskCharIndex += 1;
            } else if (maskChar.isPrepopulate()) {
                builder.append(maskChar.processCharacter(stringCharacter));
                maskCharIndex += 1;
            } else {
                strIndex += 1;
            }
        }

        return builder.toString();
    }



}

File: MaskedEditText/src/main/java/com/vicmikhailau/maskededittext/IFormattedString.kt
package com.vicmikhailau.maskededittext


interface IFormattedString : CharSequence {
    val unMaskedString: String
}


internal abstract class AbstractFormattedString(val mMask: Mask, val inputString: String) : IFormattedString {
    private var mFormattedString: String? = null
    final override val unMaskedString: String


    init {
        unMaskedString = this.buildRawString(inputString)
    }


    internal abstract fun formatString(): String

    internal abstract fun buildRawString(str: String): String


    override val length: Int
        get() = toString().length

    override fun toString(): String {
        if (mFormattedString == null) {
            mFormattedString = formatString()
        }
        return mFormattedString!!
    }

    override fun subSequence(startIndex: Int, endIndex: Int): CharSequence {
        return toString().subSequence(startIndex, endIndex)
    }

    override fun get(index: Int): Char {
        return toString()[index]
    }
}

internal class FormattedString(mask: Mask, rawString: String) : AbstractFormattedString(mask, rawString) {

    override fun buildRawString(str: String): String {
        val builder = StringBuilder()
        val inputLen = mMask.size().coerceAtMost(str.length)
        for (i in 0 until inputLen) {
            val ch = str[i]
            if (!mMask.isValidPrepopulateCharacter(ch, i))
                builder.append(ch)
        }
        return builder.toString()
    }

    override fun formatString(): String {
        val builder = StringBuilder()

        var strIndex = 0
        var maskCharIndex = 0
        var stringCharacter: Char

        while (strIndex < inputString.length && maskCharIndex < mMask.size()) {
            val maskChar = mMask[maskCharIndex]

            stringCharacter = inputString[strIndex]

            when {
                maskChar.isValidCharacter(stringCharacter) -> {
                    builder.append(maskChar.processCharacter(stringCharacter))
                    strIndex += 1
                    maskCharIndex += 1
                }
                maskChar.isPrepopulate -> {
                    builder.append(maskChar.processCharacter(stringCharacter))
                    maskCharIndex += 1
                }
                else -> strIndex += 1
            }
        }

        return builder.toString()
    }
}